# Coding Exercises (Part 1)

Now, you will have the opportunity to analyze your own dataset. <br>
__Follow the instructions__ and insert your code! You are either requested to 
- Complete the Code and __Fill in the gaps__. Gaps are marked with "__---__" and are __placeholders__ for your code fragment. 
- Write Code completely __on your own__ 

In some exercises, you will find questions that can only be answered, if your code is correct and returns the right output! The correct answer is provided below your coding cell. There you can check whether your code is correct.

If you need a hint, check the __Hints Section__ at the end of this Notebook. Exercises and Hints are numerated accordingly.

If you need some further help or if you want to check your code, you can also watch the __solutions videos__ or check the __solutions notebook__.

### Have Fun!

--------------------------------------------------------------------------------------------------------------

## Pandas Basics (DataFrame Basics I)

### Exercise 2: Indexing and Slicing

7. __Import Pandas__ (as pd), <br>
__Import__ the Cars Dataset (__cars.csv__) and save the DataFrame in the variable __cars__ and<br>
__Inspect/Print__ cars!

In [1]:
import pandas as pd

In [2]:
cars = pd.read_csv("cars.csv")

In [3]:
cars

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0    18.0          8         307.0       130.0    3504          12.0   
1    15.0          8         350.0       165.0    3693          11.5   
2    18.0          8         318.0       150.0    3436          11.0   
3    16.0          8         304.0       150.0    3433          12.0   
4    17.0          8         302.0       140.0    3449          10.5   
5    15.0          8         429.0       198.0    4341          10.0   
6    14.0          8         454.0       220.0    4354           9.0   
7    14.0          8         440.0       215.0    4312           8.5   
8    14.0          8         455.0       225.0    4425          10.0   
9    15.0          8         390.0       190.0    3850           8.5   
10   15.0          8         383.0       170.0    3563          10.0   
11   14.0          8         340.0       160.0    3609           8.0   
12   15.0          8         400.0       150.0    3761           9.5   
13   14.0          8         455.0       225.0    3086          10.0   
14   24.0          4         113.0        95.0    2372          15.0   
15   22.0          6         198.0        95.0    2833          15.5   
16   18.0          6         199.0        97.0    2774          15.5   
17   21.0          6         200.0        85.0    2587          16.0   
18   27.0          4          97.0        88.0    2130          14.5   
19   26.0          4          97.0        46.0    1835          20.5   
20   25.0          4         110.0        87.0    2672          17.5   
21   24.0          4         107.0        90.0    2430          14.5   
22   25.0          4         104.0        95.0    2375          17.5   
23   26.0          4         121.0       113.0    2234          12.5   
24   21.0          6         199.0        90.0    2648          15.0   
25   10.0          8         360.0       215.0    4615          14.0   
26   10.0          8         307.0       200.0    4376          15.0   
27   11.0          8         318.0       210.0    4382          13.5   
28    9.0          8         304.0       193.0    4732          18.5   
29   27.0          4          97.0        88.0    2130          14.5   
..    ...        ...           ...         ...     ...           ...   
368  27.0          4         112.0        88.0    2640          18.6   
369  34.0          4         112.0        88.0    2395          18.0   
370  31.0          4         112.0        85.0    2575          16.2   
371  29.0          4         135.0        84.0    2525          16.0   
372  27.0          4         151.0        90.0    2735          18.0   
373  24.0          4         140.0        92.0    2865          16.4   
374  23.0          4         151.0         NaN    3035          20.5   
375  36.0          4         105.0        74.0    1980          15.3   
376  37.0          4          91.0        68.0    2025          18.2   
377  31.0          4          91.0        68.0    1970          17.6   
378  38.0          4         105.0        63.0    2125          14.7   
379  36.0          4          98.0        70.0    2125          17.3   
380  36.0          4         120.0        88.0    2160          14.5   
381  36.0          4         107.0        75.0    2205          14.5   
382  34.0          4         108.0        70.0    2245          16.9   
383  38.0          4          91.0        67.0    1965          15.0   
384  32.0          4          91.0        67.0    1965          15.7   
385  38.0          4          91.0        67.0    1995          16.2   
386  25.0          6         181.0       110.0    2945          16.4   
387  38.0          6         262.0        85.0    3015          17.0   
388  26.0          4         156.0        92.0    2585          14.5   
389  22.0          6         232.0       112.0    2835          14.7   
390  32.0          4         144.0        96.0    2665          13.9   
391  36.0          4         135.0        84.0

8. Select the column __"name"__ with the __attribute__ (dot) notation! What is the __name__ of the car with __index label 393__?

In [4]:
cars.name

0              chevrolet chevelle malibu
1                      buick skylark 320
2                     plymouth satellite
3                          amc rebel sst
4                            ford torino
5                       ford galaxie 500
6                       chevrolet impala
7                      plymouth fury iii
8                       pontiac catalina
9                     amc ambassador dpl
10                   dodge challenger se
11                    plymouth 'cuda 340
12                 chevrolet monte carlo
13               buick estate wagon (sw)
14                 toyota corona mark ii
15                       plymouth duster
16                            amc hornet
17                         ford maverick
18                          datsun pl510
19                  vw 1131 deluxe sedan
20                           peugeot 504
21                           audi 100 ls
22                              saab 99e
23                              bmw 2002
24              

The name is ... ford mustang gl!

9. Apply also the __second alternative__ to select the column __"name"__!

In [5]:
cars["name"]

0              chevrolet chevelle malibu
1                      buick skylark 320
2                     plymouth satellite
3                          amc rebel sst
4                            ford torino
5                       ford galaxie 500
6                       chevrolet impala
7                      plymouth fury iii
8                       pontiac catalina
9                     amc ambassador dpl
10                   dodge challenger se
11                    plymouth 'cuda 340
12                 chevrolet monte carlo
13               buick estate wagon (sw)
14                 toyota corona mark ii
15                       plymouth duster
16                            amc hornet
17                         ford maverick
18                          datsun pl510
19                  vw 1131 deluxe sedan
20                           peugeot 504
21                           audi 100 ls
22                              saab 99e
23                              bmw 2002
24              

10. __Select__ the car/row at __index position 100__ with the __iloc__ operator (all columns)! How many __cylinders__ does the car have?

In [6]:
cars.iloc[100]

mpg                        18
cylinders                   6
displacement              250
horsepower                 88
weight                   3021
acceleration             16.5
model_year                 73
origin                    usa
name            ford maverick
Name: 100, dtype: object

The car has ... 6 cylinders!

11. Select the car/row at __index position 200__ with the __iloc__ operator (and only the column __"name"__)! <br>
What is the __name__ of the car?

In [7]:
cars.iloc[200, 8]

'ford granada ghia'

It´s a ... ford granada ghia!

12. Select the __last 10 rows__ and the columns __mpg, horsepower, origin__ and __name__ with the __iloc__ operator! __Fill in the Gaps__! <br>
What is the __mpg__ of the __vw pickup__?

In [8]:
cars.iloc[-10:, [0, 3, 7, 8]]

mpg  horsepower  origin                        name
388  26.0        92.0     usa  chrysler lebaron medallion
389  22.0       112.0     usa              ford granada l
390  32.0        96.0   japan            toyota celica gt
391  36.0        84.0     usa           dodge charger 2.2
392  27.0        90.0     usa            chevrolet camaro
393  27.0        86.0     usa             ford mustang gl
394  44.0        52.0  europe                   vw pickup
395  32.0        84.0     usa               dodge rampage
396  28.0        79.0     usa                 ford ranger
397  31.0        82.0     usa              chevrolet s-10

The mpg is ... 44.0!

13. __Import__ the __Cars Dataset__ one more time (save in the variable __cars__) and set the column __"name"__ as the __index__! <br>
__Fill in the Gaps__!

In [9]:
cars = pd.read_csv("cars.csv", index_col = "name")

In [10]:
cars

mpg  cylinders  displacement  horsepower  \
name                                                                           
chevrolet chevelle malibu          18.0          8         307.0       130.0   
buick skylark 320                  15.0          8         350.0       165.0   
plymouth satellite                 18.0          8         318.0       150.0   
amc rebel sst                      16.0          8         304.0       150.0   
ford torino                        17.0          8         302.0       140.0   
ford galaxie 500                   15.0          8         429.0       198.0   
chevrolet impala                   14.0          8         454.0       220.0   
plymouth fury iii                  14.0          8         440.0       215.0   
pontiac catalina                   14.0          8         455.0       225.0   
amc ambassador dpl                 15.0          8         390.0       190.0   
dodge challenger se                15.0          8         383.0       170.0   
plymouth 'cuda 340                 14.0          8         340.0       160.0   
chevrolet monte carlo              15.0          8         400.0       150.0   
buick estate wagon (sw)            14.0          8         455.0       225.0   
toyota corona mark ii              24.0          4         113.0        95.0   
plymouth duster                    22.0          6         198.0        95.0   
amc hornet                         18.0          6         199.0        97.0   
ford maverick                      21.0          6         200.0        85.0   
datsun pl510                       27.0          4          97.0        88.0   
vw 1131 deluxe sedan               26.0          4          97.0        46.0   
peugeot 504                        25.0          4         110.0        87.0   
audi 100 ls                        24.0          4         107.0        90.0   
saab 99e                           25.0          4         104.0        95.0   
bmw 2002                           26.0          4         121.0       113.0   
amc gremlin                        21.0          6         199.0        90.0   
ford f250                          10.0          8         360.0       215.0   
chevrolet c20                      10.0          8         307.0       200.0   
dodge d200                         11.0          8         318.0       210.0   
ih 1200d                            9.0          8         304.0       193.0   
datsun pl510                       27.0          4          97.0        88.0   
...                                 ...        ...           ...         ...   
chevrolet cavalier wagon           27.0          4         112.0        88.0   
chevrolet cavalier 2-door          34.0          4         112.0        88.0   
pontiac j2000 se hatchback         31.0          4         112.0        85.0   
dodge aries se                     29.0          4         135.0        84.0   
pontiac phoenix                    27.0          4         151.0        90.0   
ford fairmont futura               24.0          4         140.0        92.0   
amc concord dl                     23.0          4         151.0         NaN   
vw rabbit l                        36.0          4         105.0        74.0   
mazda glc custom l                 37.0          4          91.0        68.0   
mazda glc custom                   31.0          4          91.0        68.0   
plymouth horizon miser             38.0          4         105.0        63.0   
mercury lynx l                     36.0          4          98.0        70.0   
nissan stanza xe                   36.0          4         120.0        88.0   
honda accord                       36.0          4         107.0        75.0   
toyota corolla                     34.0          4         108.0        70.0   
honda civic                        38.0          4          91.0        67.0   
honda civic (auto)                 32.0          4          91.0        67.0   
datsun 310 gx                      3

In [11]:
cars.info()

<class 'pandas.core.frame.DataFrame'>
Index: 398 entries, chevrolet chevelle malibu to chevrolet s-10
Data columns (total 8 columns):
mpg             398 non-null float64
cylinders       398 non-null int64
displacement    398 non-null float64
horsepower      392 non-null float64
weight          398 non-null int64
acceleration    398 non-null float64
model_year      398 non-null int64
origin          398 non-null object
dtypes: float64(4), int64(3), object(1)
memory usage: 28.0+ KB


14. __Inspect__ the new Index of the DataFrame by calling the __index__ attribute!

In [12]:
cars.index

Index(['chevrolet chevelle malibu', 'buick skylark 320', 'plymouth satellite',
       'amc rebel sst', 'ford torino', 'ford galaxie 500', 'chevrolet impala',
       'plymouth fury iii', 'pontiac catalina', 'amc ambassador dpl',
       ...
       'chrysler lebaron medallion', 'ford granada l', 'toyota celica gt',
       'dodge charger 2.2', 'chevrolet camaro', 'ford mustang gl', 'vw pickup',
       'dodge rampage', 'ford ranger', 'chevrolet s-10'],
      dtype='object', name='name', length=398)

15. Select the __row__ with the index label __"ford ranger"__ (use the __loc__ operator)! What is the __weight__?

In [13]:
cars.loc["ford ranger"]

mpg               28
cylinders          4
displacement     120
horsepower        79
weight          2625
acceleration    18.6
model_year        82
origin           usa
Name: ford ranger, dtype: object

The weight is ... 2625!

16. Select the __row__ with the index label __"ford torino"__ and only the columns __"model_year"__ and __"origin"__! <br>
What´s the __model year__?

In [14]:
cars.loc["ford torino", ["model_year", "origin"]]

model_year     70
origin        usa
Name: ford torino, dtype: object

The model year is ... 70!

17. Select the columns __"mpg"__ and __"weight"__ (all rows) with the __loc__ operator!

In [ ]:
cars.loc[:, ["mpg", "weight"]]

# Well Done!

---------------------------------------------------------------------------------------------------------------------

# Hints (Spoiler!)

7. pandas as pd, pd.read_csv("filename"), head()

8. cars.---

9. cars["---"]

10. cars.iloc[---]

11. cars.iloc[row index position, column index position]; "name" column is at index position ... 8

12. The last 10 rows = from index position -10 till the end (:); Count the column index positions of the Columns


13. parameter index_col

14. cars.---

15. cars.loc["row label"]

16. cars.loc[---, [---, ---]]

17.  cars.loc[ : , [---, ---]]